pkg/services/dashboards/service/client/client.go

```diff
@@ -91,6 +128,144 @@ func (h *K8sClientWithFallback) Get(ctx context.Context, name string, orgID int6
+func (h *K8sClientWithFallback) List(
+	ctx context.Context, orgID int64, options metav1.ListOptions,
+) (*unstructured.UnstructuredList, error) {
+	ctx, span := tracing.Start(ctx, "K8sClientWithFallback.List")
+	defer span.End()
+
+	span.SetAttributes(
+		attribute.Int64("org.id", orgID),
+		attribute.Bool("fallback", false),
+	)
+
+	span.AddEvent("List")
+	initial, err := h.K8sHandler.List(ctx, orgID, options)
+	if err != nil {
+		h.log.Error("failed to fetch initial list", "error", err)
+		return nil, tracing.Error(span, err)
+	}
+
+	res := initial.DeepCopy()
+	res.Items = res.Items[:0]
+
+	// Map of version -> list of names to fetch.
+	toFetch := make(map[string][]nameAndResourceVersion)
+
+	for _, item := range initial.Items {
+		failed, storedVersion, conversionErr := getConversionStatus(&item)
+		if !failed {
+			res.Items = append(res.Items, item)
+			continue
+		}
+
+		h.log.Debug(
+			"will fetch object with the stored version",
+			"name", item.GetName(),
+			"storedVersion", storedVersion,
+			"conversionErr", conversionErr,
+		)
+		h.metrics.fallbackCounter.WithLabelValues(storedVersion).Inc()
+
+		names, ok := toFetch[storedVersion]
+		if !ok {
+			names = make([]nameAndResourceVersion, 0)
+		}
+
+		names = append(names, nameAndResourceVersion{
+			Name:            item.GetName(),
+			ResourceVersion: item.GetResourceVersion(),
+		})
+		toFetch[storedVersion] = names
+	}
+
+	for version, names := range toFetch {
+		h.log.Info(
+			"fetching multiple objects with the stored version",
+			"version", version,
+		)
+
+		// Log names at debug level, because there could be a lot of them.
+		h.log.Debug(
+			"will fetch objects with names",
+			"names", names,
+		)
+
+		items, err := h.fetchWithVersion(ctx, orgID, version, names...)
+		if err != nil {
+			return nil, tracing.Error(span, err)
+		}
+
+		res.Items = append(res.Items, items...)
+	}
+
+	return res, nil
+}
+
+// fetchWithVersion fetches multiple resources from the K8s API.
+// It uses concurrent Get requests, one for each name.
+//
+// TODO: maybe consider using List with a field selector and / or search?
+//
+//	items, err := client.List(spanCtx, orgID, metav1.ListOptions{
+//		FieldSelector: fmt.Sprintf("metadata.name in (%s)", strings.Join(names, ",")),
+//	})
+func (h *K8sClientWithFallback) fetchWithVersion(
+	ctx context.Context, orgID int64, version string, items ...nameAndResourceVersion,
+) ([]unstructured.Unstructured, error) {
+	ctx, span := tracing.Start(ctx, "K8sClientWithFallback.fetchWithVersion")
+	defer span.End()
+
+	span.SetAttributes(
+		attribute.String("version", version),
+		attribute.Int("count", len(items)),
+	)
+
+	client := h.newClientFunc(ctx, version)
+
+	g, ctx := errgroup.WithContext(ctx)
+	g.SetLimit(maxConcurrentGetRequests)
+	res := make([]unstructured.Unstructured, len(items))
+
+	for i, it := range items {
+		g.Go(func() error {
+			ctx, span := tracing.Start(ctx, "K8sClientWithFallback.fetchWithVersion.ConcurrentGet")
```

Comment:
In case you agree with comment above:
```suggestion
			ctx, span := tracing.Start(errCtx, "K8sClientWithFallback.fetchWithVersion.ConcurrentGet")
```
I'm ok with leaving `ctx, span` here as I think the scope is short enough to not make this more complicated to understand at a first sight.